File: PyWagg/test/parser_test.py
import unittest
from PyWagg import tokens
from PyWagg import lexer
from PyWagg import ast
from PyWagg import parser


class ParserTest(unittest.TestCase):

    def test_let_statements(self):
        tests = [
            ("boop x = 5;", "x", 5),
            ("boop y = 10;", "y", 10),
            ("boop foobar = 838383;", "foobar", 838383),
        ]

        for t in tests:
            l = lexer.new(t[0])
            p = parser.new(l)
            program = p.parse_program()
            self.check_parse_errors(p)
            stmt = program.statements[0]
            if not self.check_let_statement(stmt, t[1]):
                return

    def check_parse_errors(self, p):
        errors = p.errors
        if len(errors) == 0:
            return
        print('parser has {} errors'.format(len(errors)))
        for e in errors:
            print('parser error: {}'.format(e))
        self.fail()

    def check_let_statement(self, s, name):
        if s.token_literal() != 'boop':
            print("s.token_literal not 'boop'. got={}".format(s.token_literal()))
            return False
        if not isinstance(s, ast.LetStatement):
            print("s is not a ast.LetStatement. got={}".format(type(s)))
            return False
        if s.name.value != name:
            print("statement s value is not {}. got={}".format(name, s.name.value))
            return False
        if s.name.token_literal() != name:
            print("statement s token is not {}. got={}".format(name, s.name.token_literal()))
            return False
        return True


if __name__ == '__main__':
    unittest.main()

File: GoWagg/parser/parser_test.go
package parser

import (
	"Wagg/GoWagg/ast"
	"Wagg/GoWagg/lexer"
	"testing"
)

func TestLetStatements(t *testing.T) {
	input := `
	boop x = 5;
	boop y = 10;
	boop foobar = 838383;
	`

	l := lexer.New(input)
	p := New(l)

	program := p.ParseProgram()
	checkParserErrors(t, p)
	if program == nil {
		t.Fatal("ParseProgram() returned nil")
	}

	if len(program.Statements) != 3 {
		t.Fatalf("program.Statements does not contain 3 statements. Got =%d",
			len(program.Statements))
	}

	tests := []struct {
		expectedIdentifier string
	}{
		{"x"},
		{"y"},
		{"foobar"},
	}

	for i, tt := range tests {
		stmt := program.Statements[i]
		if !testLetStatement(t, stmt, tt.expectedIdentifier) {
			return
		}
	}
}

func testLetStatement(t *testing.T, s ast.Statement, name string) bool {
	if s.TokenLiteral() != "boop" {
		t.Errorf("s.TokenLiteral not 'boop'. Got =%q", s.TokenLiteral())
		return false
	}

	letStmt, ok := s.(*ast.LetStatement)
	if !ok {
		t.Errorf("S not *ast.LetStatement. Got =%s", s)
		return false
	}

	if letStmt.Name.Value != name {
		t.Errorf("letStmt.Name.Value not '%s'. Got =%s", name, letStmt.Name.Value)
		return false
	}

	if letStmt.Name.TokenLiteral() != name {
		t.Errorf("letStmt.Name.TokenLiteral() not '%s'. Got =%s",
			name, letStmt.Name.TokenLiteral())
		return false
	}

	return true
}

func checkParserErrors(t *testing.T, p *Parser) {
	errors := p.Errors()
	if len(errors) == 0 {
		return
	}

	t.Errorf("Parsers has %d errors", len(errors))
	for _, msg := range errors {
		t.Errorf("Parser error: %q", msg)
	}
	t.FailNow()
}

func TestReturnStatements(t *testing.T) {
	input := `
	bork 5;
	bork 10;
	bork 993322;
	`

	l := lexer.New(input)
	p := New(l)

	program := p.ParseProgram()
	checkParserErrors(t, p)

	if len(program.Statements) != 3 {
		t.Fatalf("program.Statements does not contain 3 statements. Got =%d",
			len(program.Statements))
	}

	for _, stmt := range program.Statements {
		returnStmt, ok := stmt.(*ast.ReturnStatement)
		if !ok {
			t.Errorf("stmt not *ast.ReturnStatement. Got=%T",
				stmt)
			continue
		}
		if returnStmt.TokenLiteral() != "bork" {
			t.Errorf("returnStmt.TokenLiteral not 'bork', got=%q",
				returnStmt.TokenLiteral())
		}
	}
}
